pkg/core/native/contract.go

```diff
@@ -1,28 +1,100 @@
 	"github.com/nspcc-dev/neo-go/pkg/core/interop"
 	"github.com/nspcc-dev/neo-go/pkg/core/interop/interopnames"
+	"github.com/nspcc-dev/neo-go/pkg/core/native/noderoles"
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
 	"github.com/nspcc-dev/neo-go/pkg/io"
 	"github.com/nspcc-dev/neo-go/pkg/util"
 	"github.com/nspcc-dev/neo-go/pkg/vm/emit"
 )
 
+// Native contract interfaces sufficient for Blockchain functioning.
+type (
+	IManagement interface {
+		interop.Contract
+		GetNEP11Contracts(d *dao.Simple) []util.Uint160
+		GetNEP17Contracts(d *dao.Simple) []util.Uint160
+	}
+	ILedger interface {
+		interop.Contract
+	}
+	INEO interface {
+		interop.Contract
+		GetCommitteeAddress(d *dao.Simple) util.Uint160
+		GetNextBlockValidatorsInternal(d *dao.Simple) keys.PublicKeys
+		BalanceOf(d *dao.Simple, acc util.Uint160) (*big.Int, uint32)
+		CalculateBonus(ic *interop.Context, acc util.Uint160, endHeight uint32) (*big.Int, error)
+		GetCommitteeMembers(d *dao.Simple) keys.PublicKeys
+		ComputeNextBlockValidators(d *dao.Simple) keys.PublicKeys
+		GetCandidates(d *dao.Simple) ([]state.Validator, error)
+	}
+	IGAS interface {
+		interop.Contract
+		BalanceOf(d *dao.Simple, acc util.Uint160) *big.Int
+	}
+	IPolicy interface {
+		interop.Contract
+		GetStoragePriceInternal(d *dao.Simple) int64
+		GetMaxVerificationGas(d *dao.Simple) int64
+		GetExecFeeFactorInternal(d *dao.Simple) int64
+		GetMaxTraceableBlocksInternal(d *dao.Simple) uint32
+		GetMillisecondsPerBlockInternal(d *dao.Simple) uint32
+		GetMaxValidUntilBlockIncrementFromCache(d *dao.Simple) uint32
+		GetAttributeFeeInternal(d *dao.Simple, attrType transaction.AttrType) int64
+		CheckPolicy(d *dao.Simple, tx *transaction.Transaction) error
+		GetFeePerByteInternal(d *dao.Simple) int64
+	}
+	IOracle interface {
+		interop.Contract
+		GetOracleResponseScript() []byte
+		GetRequests(d *dao.Simple) (map[uint64]*state.OracleRequest, error)
+		GetScriptHash(d *dao.Simple) (util.Uint160, error) // TODO: this method should be removed, Designation is sufficient for that
+		GetRequestInternal(d *dao.Simple, id uint64) (*state.OracleRequest, error)
+		SetService(o OracleService)
+	}
+	IDesignate interface {
+		interop.Contract
+		GetDesignatedByRole(d *dao.Simple, r noderoles.Role, index uint32) (keys.PublicKeys, uint32, error)
+		GetLastDesignatedHash(d *dao.Simple, r noderoles.Role) (util.Uint160, error)
+		SetOracleService(o OracleService)
+		SetNotaryService(n NotaryService)
+		SetStateRootService(s StateRootService)
+	}
+	INotary interface {
+		interop.Contract
+		BalanceOf(dao *dao.Simple, acc util.Uint160) *big.Int
+		ExpirationOf(dao *dao.Simple, acc util.Uint160) uint32
+		GetMaxNotValidBeforeDelta(dao *dao.Simple) uint32
+	}
+	ICrypto interface {
+		interop.Contract
+	}
+	IStd interface {
+		interop.Contract
+	}
+)
+
 // Contracts is a set of registered native contracts.
 type Contracts struct {
```

Comment:
I'd try leaving this struct as is, rather adapting it's main (and the only?) user in core. Many contracts it doesn't even care about. You can have some methods defined on it to get neo/policy/whatever else and then Blockchain would store and use these interfaces instead of `Contracts` structure, while `Contracts` would return proper `*NEO`, `*Policy` and whatever else without thinking much about interfaces.